bigquery/schema.go

```diff
@@ -533,6 +542,11 @@ func inferFieldSchema(fieldName string, rt reflect.Type, nullable bool) (*FieldS
 		return &FieldSchema{Required: !nullable, Type: BooleanFieldType}, nil
 	case reflect.Float32, reflect.Float64:
 		return &FieldSchema{Required: !nullable, Type: FloatFieldType}, nil
+	case reflect.Map:
```

Comment:
I wonder if we should treat a `reflect.Map`, the same way we treat `reflect.Struct` - we map it to BigQuery Struct/Records as check it recursively for such conversion. And only when the `json` tag is there we map it to a BQ JSON type. The reason I'm putting this is that if we don't do it now, later if we decide to infer a map as a BQ Struct/Record, we will either cause a breaking change or we would have to introduce something like a `record` tag to have that behavior, which can be confusing.

Reply:
Can I define a column that's a `STRUCT` without its fields being known ahead of time? The main case I want to support is a grab-bag of unknown-at-schema-time fields which may come in from a user's request (think HTTP request headers).

Today I maintain an allowlist of known/expected headers, but ideally I'd be able to toss any header I get into the schema, and later query `SELECT * FROM records WHERE headers.user-agent == "blah"` without having to define `user-agent`.

If I can do that with a schema-less `STRUCT` I'm okay with that too.

Reply:
BQ has the concept of anonymous structs, which are typed but are not keyed with explicit names.  The late-defined example here is definitely best matched to the semistructured capabilities of the JSON type.

If BQ introduces first class map support there's a lot of expectations that need to be revisited up and down the stack, this is but one of many facets of that.